### DISCUSSION TITLE
Pymc3 Installation Issues in Linux

### DISCUSSION THREAD

**Discussion by Hyunglok-Kim** [pymc-devs/pymc3#6725] [Q&A]
Hi,
I am trying to install pymc3 using the following command line in Linux:

conda create -c conda-forge -n pymc3_env pymc3 theano-pymc mkl mkl-service

However, I got the following error when I try to import pymc3, and I suspect this is related to the numpy version:

>>> import pymc3
/home/h/anaconda3/envs/pymc3/lib/python3.10/site-packages/theano/scalar/basic.py:2412: FutureWarning: In the future `np.bool` will be defined as the corresponding NumPy scalar.
  self.ctor = getattr(np, o_type.dtype)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/h/anaconda3/envs/pymc3/lib/python3.10/site-packages/pymc3/__init__.py", line 23, in <module>
    import theano
  File "/home/h/anaconda3/envs/pymc3/lib/python3.10/site-packages/theano/__init__.py", line 83, in <module>
    from theano import scalar, tensor
  File "/home/h/anaconda3/envs/pymc3/lib/python3.10/site-packages/theano/scalar/__init__.py", line 1, in <module>
    from .basic import *
  File "/home/h/anaconda3/envs/pymc3/lib/python3.10/site-packages/theano/scalar/basic.py", line 2460, in <module>
    convert_to_bool = Cast(bool, name="convert_to_bool")
  File "/home/h/anaconda3/envs/pymc3/lib/python3.10/site-packages/theano/scalar/basic.py", line 2412, in __init__
    self.ctor = getattr(np, o_type.dtype)
  File "/home/h/anaconda3/envs/pymc3/lib/python3.10/site-packages/numpy/__init__.py", line 305, in __getattr__
    raise AttributeError(__former_attrs__[attr])
AttributeError: module 'numpy' has no attribute 'bool'.
`np.bool` was a deprecated alias for the builtin `bool`. To avoid this error in existing code, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
The aliases was originally deprecated in NumPy 1.20; for more details and guidance see the original release note at:
    https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations. Did you mean: 'bool_'?

Could you please provide me with any advise on this issue?
Thank you.

**Comment by twiecki** [comment]
Any reason you can't use PyMC 5.x?

**Reply by Hyunglok-Kim** [reply]
Hi, 

Yes, I am attempting to utilize the continuous zero-inflated exponential likelihood distribution using pm.DensityDist with:

```
def  logp_zero_inflated_exponential(value, psi, mu):
     return tt.switch(tt.eq(value, 0), tt.log(psi), tt.log1p(-psi) - mu * value)

psi = pm.Beta('psi', 1, 1)
z  = pm.HalfNormal('Z', sigma=500)
a  = pm.HalfNormal('α', sigma=1000) # weak informative prior (we know it should be positive) 
b  = pm.Normal('β', mu=0, sigma=1000)

ET_I_est = f_ET_I(SM1, SM2, a, b, dt)
mu = z*np.abs(dsm) - ET_I_est

Y_obs = pm.DensityDist('Y_obs', logp_zero_inflated_exponential, observed={'value': y_obs, 'psi': psi, 'mu': mu})
```


This is working fine with the version 3, however, when I use it with 
`Y_obs = pm.DensityDist('Y_obs', logp_zero_inflated_exponential, observed=np.array([y_obs, psi, mu]))`
in the version 5, I got the following error:

```
    860                 Y_obs = pm.DensityDist('Y_obs', logp_zero_inflated_exponential, 
    861                                        observed={'value': y_obs, 'psi': psi, 'mu': mu})
    862             elif pmv == 5:
    863                 Y_obs = pm.DensityDist('Y_obs', logp_zero_inflated_exponential, 
--> 864                                        observed=np.array([y_obs, psi, mu]))
    865 # sampling or fit
    866 rng = 321

ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (3,) + inhomogeneous part.
```

I might be missing something with the Version 5 API. If you can provide me with instructions on how to use the Version 5 API correctly, that would be really helpful. If it works, I won't need to use pymc3. By the way, the conda-forge pymc3 library installed on macOS works just fine with the mkl-service library. I'm not sure why the conda-forge pymc3 library for Linux comes with a numpy version that is not 1.22.

Thank you for your help.

**Reply by twiecki** [reply]
@Hyunglok-Kim So I would try and get it to work on PyMC 5.x. As to the question, can you post that on discourse.pymc.io?

**Reply by Hyunglok-Kim** [reply]
Sure thing. I have posted it on discourse.pymc.io. Check [here](https://discourse.pymc.io/t/pm-densitydist-issue-pymc3-vs-pymc5/12181).

**Reply by Hyunglok-Kim** [reply]
Additional information:

I use different logp_zero_inflated_exponential function for pymc3 (pmv == 3) and pymc5:
```
if pmv == 3:
    def logp_zero_inflated_exponential(value, psi, mu):
        return tt.switch(tt.eq(value, 0), tt.log(psi), tt.log1p(-psi) - mu * value)

if pmv == 5:
    def logp_zero_inflated_exponential(value, psi, mu):
      if value == 0:
        return pm.math.log(psi)
      else:
        return pm.math.log1p(-psi) - mu * value
```
